### DISCUSSION TITLE
How to prevent refetching of component data on every route change

### DISCUSSION THREAD

**Discussion by milelazar** [nuxt/nuxt#19912] [Questions] [answered]
I have a dynamic [...slug] page, and Header component with useFetch API call.
And when navigating around, fetch call in the Header component is executed for every route change. That's not what I want.

I need Header (and any subcomponent) to fetch data only once, and remain static after fully loaded the first time. Not fetching the same data on every route change. I guess I could create a state for each fetch call and use that if data is already fetched. But that doesn't seem right to me. Is there another way?

Here is a simple reproduction: https://stackblitz.com/edit/nuxt-starter-ia6qzh?file=components/Header.vue

**Comment by TheAlexLichter** [comment, marked as answer]
You could use a layout or put the header in your `app.vue`. Alternatively, you could also build yourself a little caching logic, e.g. via `useAsyncData`.

**Reply by milelazar** [reply]
Thanks for the suggestions.

I have already tried with layouts but there is no difference. Header data still gets refetched every time. 
App.vue solution without layouts works. But i need layouts to be able to override slots in some templates.

Here is an example of what I actually want to achieve: https://stackblitz.com/edit/nuxt-starter-ezoqnu?file=pages/[...slug].vue

There is a catch all page with dynamic component. Dynamic component is then loading global subtemplates based on another API response (in the example it is hardcoded to do it based on route path).

Looks like I'll try the caching method as you suggested

**Reply by TheAlexLichter** [reply]
Sounds like a plan!

See e.g. something like https://stackblitz.com/edit/github-vcbsj4?file=app.vue ([src](https://twitter.com/henrycunh/status/1630671452239196160)).

**Reply by milelazar** [reply]
I forgot to check useNuxtData. It looks like a simple and effective solution. I'll try with that. Thanks for your help!
